### DISCUSSION TITLE
Strategies to complete map matching on incomplete data

### DISCUSSION THREAD

**Discussion by DigNZ** [valhalla/valhalla#3668] [Q&A] [answered]
I have traces for a day's driving from telematics devices in cars and I need to snap them to roads in order to work on further data processing. My system uses valhalla for routing and attributes so it makes sense to use it for map matching routes as well.

BUT it seems to be an all or nothing kind of thing and when the route doesn't work because of some sections then I get nothing back for any of the route and no specific indication on where the problems could be. Some examples of the data I think is breaking trace_route:
![image](https://user-images.githubusercontent.com/817657/176139214-007aa382-4226-4eb6-8953-c483af1c33da.png)
![image2](https://user-images.githubusercontent.com/817657/176139328-0361f5b7-f444-41c0-aa1a-098e18238a24.png)
![image3](https://user-images.githubusercontent.com/817657/176139349-45699117-f1a4-4791-abed-0fdd075a1649.png)

I've done a lot of reading about how to help things and nothing much seems to help. forcing shape_match to map_snap (and not rely on walk_or_snap) returns data but it can be wildly inaccurate for the rest of the route and often shows up in the alternates only with the main trip being very short (presumably because the start and end points are next to each other).

ANYWAYs what are some strategies I could use for getting as much as possible to match even if it involves discarding sections that can't match. I tend to get errors like:
`{
    "error_code": 299,
    "error": "Unknown:Could not build directions for TripLeg",
    "status_code": 400,
    "status": "Bad Request"
}`
Unless the trip closely follows roads at all points throughout in which case it all works perfectly.

I've seen mention of tweaking ignore_ options but can't find what they are or what to try. I don't mind if long sections skipped and can probably identify them in my data before sending it through but short sections though building and carparks need to be identified and matched or skipped to prevent the rest of the trace from failing.

**Comment by DigNZ** [comment]
Not sure if this is clearer but here is a link to a polyline that fails trace_route:

If I pull out enough points then it will eventually work. I believe its the corner that's cut through buildings in the lower left of the map that's causing the issue but I don't know for sure. Looking for a strategy to manage this.

Thanks!
[Example Polyline](https://valhalla.github.io/demos/polyline/?unescape=false&polyline6=true#z%7BaymAshfjlI%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3FcoAehE%3FyBcBwDsP%3FkU%3FyC%3Fs%40%5C%3F%3F%3F%3F%3F%3Fr%40%5Ds%40%5Cu%40%3FwCuD%3F%7DUr%40k%60%40t%40%7Bi%40%3F%7Bj%40eJiMcG%3FiM%7D%40iR%3FaB%5Du%40%3FqHyB%3FkK%3FaRt%40_Tl%40ku%40t%40gb%40%3FqHcBwCmE%7D%40iR%5Ds%60%40%7D%40ab%40%7B%40%7DN%5DcG%7D%40u%40%3Fm%40%3FiR%3FmE%7C%40ssAxVse%40lJaB%5CmE%7C%40cB%3FcGzAmEz%40aB%3FcB%7C%40s%40%3F%3F%3F%3F%3F%3F%3F%3F%3FkB%5C%3F%3FaB%7C%40%3F%7D%40u%40%3FwC%7BAmEq%5Bm%40%7DUyHqp%40wC%7BVu%40mJ%7BFun%40mEk%60%40wCwXcBeO%3F%7BA%3F%3F%3F%3F%3FwCiBaRwCuZcBaRiB%7BU%7BF%7BtA%3F%7Dh%40%3Fo%5DvCib%40dJ%7DTl%40wCxHkLnHsFxC%7B%40lE%7BAlE%7BAvCqGzFkKpMs%5Bx%5Bq%5D~NzAtc%40zUlUsE%7CNuZaBq%5Cu%40q%5CdJePhRaRvSed%40zFmJhb%40oIzKwCdEqH~CcQbBgM%3F%7BAr%40%7D%40%3F%3F%3FuEbBgMhBcRaGoHqH%7BA%7BKyB%3F%3FwC%5DcGqHvHwXfJiMdZo%5DrKuZlEq%5BvCsFt%40%5C%3F%3Fu%40%5Dt%40%3Fr%40%5CzFxBlE%5CxC%7C%40r%40vC%3FvCs%40%3F%3F%3F%3F%7C%40r%40%3F%3F%3F%3F%5CaBpH_Dfb%40qH%7C%40t%40mJ%60Bz%40%3F%3F%3F%3F%3F%3F%3F%3Fm%40%7B%40s%40_%40%3F%3F%3F%3Fr%40%7B%40l%40uE%3F%7BAlEwXt%40%3FlEwC%3F%7D%40%3F%3F%3F%3Fl%40yAvHxAhMzBlUnHfb%40lJfJxClE%5C%3Fz%40%3F%60Ru%40lKcGvm%40oHlr%40yC%7CTs%40fO%3F%3F%3F%3F%3Fz%40u%40xCwCrFmEtE%7BFrFyCrEoMhMqHxCmE%3FsP%3FeEyCcGwCiMwCcG%3FgM%3FkB%3FwC%3FwC%3FcB%3F_TvCsK%7C%40o%5DzToXde%40wSxW%3Fj%60%40%60B~f%40kU%7CTcW%5DyVeOeZyCiBzA%7BFjLqMbe%40aRvXiRrFwCzA%3F%3F%3F%3Fu%40%3FeJnIgErFmElJaGdPcBxV%3Fju%40l%40dOt%40hM%3F%7C%40u%40%3F%3F%3F%3Fz%40t%40hM%3FtYlEhb%40rP%7CfBdJ%7C%7D%40l%40p%5CaBxCmEz%40wStE_OtEcG%5CwC%7C%40%7BFxAs%60%40lJ%7BKzAmExBmE%5CgMxCiRtEgEz%40s%40%7C%40u%40%5Ct%40%60R%60BzUxC%7CTr%40lJ%3F%7C%40t%40hMt%40jKl%40lJvCfO%3F%5C%3F%3Ft%40%5D%3FxAr%40rGrK%3FjB%3F%3F%3F%3F%3Fu%40%5Cu%40%5D%3F%3F%3F%3F%3F%7D%40%3F%3F%3F%3Ft%40vC%3F%3Ft%40%3Ft%40%7C%40dE%7D%40~C%3FdJyAvC%7D%40%3F%3Fl%40%3Ft%40%7B%40pMrEhMxC%60WrFrKvCrPvC~NxCzKvCz~%40~RdZrFzKzAt%40%7C%40u%40%3Fu%40%7D%40%3F%3Ft%40%7C%40vC%3FtPz%40dJ%3F%7C%5E%5Eje%40xBxxCrF%60R%3FhMvC~CdOu%40zV%3FxVs%40~h%40u%40p%5Bt%40pHr%40tE%3F%5C%3F%3F%3F%3Ft%40xB~N%7C%40nH%5CdJz%40t%40%3F%3F%3Fl%40%3FhB%3Fl%40%7C%40jBdPu%40j_%40%3Fde%40u%40bP%3FpHm%40lJ%3FlJbB~SvC%5C~S%7C%40rK%3FlEtEs%40lI%3F%7CUr%40lJbGvC~N%3FxF%5CjB%3F%3F%3F%3F%3F%60B%5Dt%40%7D%40%3FyA%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F%3FcoAkeE%3F%7D%40%3F%7B%40aBsFcG%7D%40%7BF%5DcG%3FmE%3Fk%40%3Fu%40%3F%3F%3F%7BKqGm%40cR%3Fq%5B%3F%7BsAiBcQmEyCwSyA_O%7D%40%7DNwCu%40%7BAu%40mJt%40_T%3FuY%3Fs%5Bt%40ag%40%3FeOkBsF%7BF%7D%40o%5D%7B%40ab%40%5DeZ%7D%40iR%7B%40%7BF_%40e_%40xCaRvCq%60A~Ryk%40jLsPzAwCz%40u%40%3FcB%3FwCyA_DaScBqH%3F%3F%3F%3F%3F%7C%40%3F%3F%3F%3Fs%40%7D%40%3F%3F%3F%3F%3F%3F%3F%3FhB%5Du%40mJaByVcG%7Bj%40sPaeBqHgw%40sPebBmE%7B%7CBhBso%40hMs%5BhByCfEsF%3F%3F%3F%7B%40%3F%3F%60GsFbB%3FvC%7D%40jB%3Fr%40%7B%40xC%5D%60B%7BBxCuDhBwCbBsGvCoHbGeP%60Ro%5DnMyCbRtEje%40dPhRuEzKgNl%40gNmEms%40dJ%7DTfr%40wbAvXsF%60RuEpHaRhBcQxCyV%3F%7BAr%40wC%3F%7D%40%3F%3F%3FwCl%40kKm%40%7D%40_D%7B%40eJyCwCyAcB%3F%3F%3F%3F%3Fu%40%7D%40%3F%3Fm%40%3F_D%5D%7BFyCs%40gNrPq%5CvS_S%60Wyl%40vCaRbBeOlEib%40vCsZtPsyA%60G%7Dh%40xCgOvCyA%60BrEjBlK%60WlJlPvCje%40lKdZnHlExBhB%3Fl%40%3FhBoHxCs%5B%60BcQbGib%40rP%7DTfm%40_g%40d_%40s%5BlUib%40dEiMbGmJlEvCnHtFhMhLz%5B%7CTfb%40bQxCvD%3FfMmEvXmEp%5DdJ%60Rm%40tEbBrFdJvCpHpG%3F%5E%3F%3F%3F%3Fm%40%3Fu%40%3F%3F%3F%3F%3F%3F%3F%3F_%40%3F%3Fu%40%7B%40%7BF%7BAiB%7B%40wC%7D%40u%40%5DcBuE%3FwCaByCmEmJ%60Bk%60%40xC%7DSr%40iM%3FyCs%40wCmUmJ_OqHkUcPiMkLeJoHmE%3F%7BK%60Qi%5D%60g%40%7Dc%40n%5Eg%5DvXiR~SgJtYsK%3F_TwDyVqGcz%40_T%7BKlJu%40rGwCnIwC%7BAr%40aRaGs%5BeZyWmUePiMsFmE%7B%40wC%3FwHzAuKz%40oMoH%7BFq%5DdJ%7Dh%40bB%7B_A_OuwA%7BKk%7DAm%40yW_DaSeJ_SqHyWkUia%40wc%40q%5Cy%5B%7DTqHqGwCiMl%40aSdJ%7Bj%40dJyk%40~CcPrK%7Bk%40tP_g%40nH%7BVjBiMu%40%3FmE%3F%7BF%5Cu%40%7C%40%3F%3Ft%40%7D%40%3F%3F%3F%3Fu%40%7C%40%3F%3F%3F%3FxCwC%3Fz%40%3F%7B%40%3F%3Fu%40%7D%40wCuD%60GuEzFnHxCzA~CwCzFaR%60GePfJsZxFiMs%40rFiMl_%40iMj%60%40eZ~%7B%40%7BKr%5BoHzUu%40nIqHha%40aGr%5B%7BFfc%40u%40tEiBhLcB%60ShMbQrPjKlEtElEvCzFvCxk%40de%40bGtEvCz%40pH%5ChR%5DdJrFpHtZqHl%5EeJhb%40lEdx%40rKvbAvHfw%40l%40p%5C%3FlJ%3FpG~Cr%5BfEpHdJ%60RmEdd%40%7BVpHwXxAq%5Bj%60%40_Tl_%40oh%40fx%40coAfuB%7Bn%40tbA__%40ls%40aGzVsKxuA_Dfb%40yCn%5DaBlK%3F%3F%3F%3Fu%40%3Fu%40xBaB%60RqHhv%40u%40nr%40s%40tn%40%3F%60g%40lEhLhB%3FpH%3FfM%5EhRz%40~NbQt%40zAl%40lJiMxV_O%60h%40u%40%60Q%3Fj%60%40t%40jL%3FfNbBhM%3F%3F%3F%3F%3FfO%3FrZlExk%40lEde%40%7CNnzAfJfx%40vCjKhB%3FlPwCdOwCxf%40mK~NwCvH%7BAfJyBdE%5DlE%7D%40rPsEdZsFvH%7D%40bBvC%3FrF%60BbQjBfOl%40nHhB%7CTvCtZbBjKhBbQt%40jKl%40zA%3F%3F%3F%3F%3F%7C%40t%40%3FnHnH~SxB%60b%40%5Efb%40z%40lU%7C%40pHtD%3FpHu%40xk%40%3Fp%5Cm%40~R%3Fl_%40%3FlKl%40xBdJxA~S%7C%40nXzAzFdOm%40be%40%3F%7Ci%40%3Fdd%40%3FfOl%40xAhBxChMxBdJ%3FfJ%5C%3F%3FhB%3Fl%40z%40%3F%3FvC%7C%40t%40vDu%40dd%40m%40p%5B%3FzVs%40%60R%3FzU%3FfOr%40xAlE%7C%40zF%5CpM%3FzF%7C%40bGvC%3FrFu%40xW%3FnHbGrFdJ%3F%7CI%7C%40%60D%3FvC%7D%40%3FsE%3F%3F%3F%3F%3F%3F%3F%3F%3F%3F)

**Reply by kevinkreiser** [reply]
@DigNZ is there any possibility you'd be able to whittle the trace down to less coordinates? its really quite a bit easier to debug with smaller input.

**Reply by DigNZ** [reply]
I'll try. I find that when I whittle it down it starts working! I'll aim to isolate part the part that is failing.

**Reply by DigNZ** [reply]
Ok, so digging into it got a little stranger. In that example there are 834 points. trace_route with them all fails, however if I do them in blocks of 100, or 200, or 300 (multiple calls with parts of the route) then it works all the way through. If I run it with an increasing number of points each time then it fails if the point at index 14 and the point at index 361 is included in the same route. 
`
{ "error_code": 299, "error": "Unknown:Could not build directions for TripLeg", "status_code": 400, "status": "Bad Request" }
`

Point 14 [ 174.771866, -41.321934 ]
Point 361 [ 174.77183499999998, -41.321369 ]

Remove either of those points and the trace_route will succeed. Seems very strange.

**Reply by DigNZ** [reply]
OK, so here is the most minimal example:
trace_route for these two points succeeds:
[Success](https://valhalla.github.io/demos/polyline/?unescape=false&polyline6=true#z%7BaymAshfjlI%7Dc%40oH)
And these two fail:
[Fails](https://valhalla.github.io/demos/polyline/?unescape=false&polyline6=true#z%7BaymAshfjlIib%40%7C%40)

trace_attrbutes returns 'matched' for both in both examples but only returns edges for the first, returns an empty array for the second one.

**Comment by kevinkreiser** [comment, marked as answer]
in your picture above it looks like you have long stretches of distance with no points along them, am i misunderstanding that? looking at the polyline link this also seems to be the case in the left most edge of the area. there is a feature called "breakage distance" this is defaulted to 2km. if the distance between any two points in the input is larger than this distance then the map matcher will automatically break the route there and not try to find a route between those two points. if this is too small of a distance for you you can change that config value: https://github.com/valhalla/valhalla/blob/master/scripts/valhalla_build_config#L130

**Reply by DigNZ** [reply]
If it breaks a route there will it then present two (which is fine) or will it stop at the break (which is less ideal)?
